### ISSUE TITLE
SFT OOM 0.7.1 -> 0.7.1 + including 0.7.7 breaking change.

### ISSUE THREAD

**Issue by ChristianPala** [huggingface/trl#1168]
Hi folks,

Happy New Year!

We noticed a breaking change to our model training code, I provided a simplified example below, any idea why it works perfectly fine up to trl 0.7.1, but breaks with the later versions?

```python
import torch
from datasets import load_dataset
from peft import LoraConfig
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig, TrainingArguments
from trl import DataCollatorForCompletionOnlyLM, SFTTrainer


def main():        
    quantization_config = BitsAndBytesConfig(
        load_in_4bit=True,
        bnb_4bit_compute_dtype=torch.bfloat16,
        )

    dataset = load_dataset("mlabonne/guanaco-llama2-1k", split="train")

    model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-2-7b-chat-hf", 
                                                quantization_config=quantization_config, 
                                                torch_dtype=torch.bfloat16)

    tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-2-7b-chat-hf")
    tokenizer.pad_token = tokenizer.eos_token


    response_template = "[/INST]" 
    collator = DataCollatorForCompletionOnlyLM(response_template, tokenizer=tokenizer)


    training_args = TrainingArguments(
        output_dir="test",
        bf16=True,
        neftune_noise_alpha= 5,        
    )

    peft_config = LoraConfig(
        r=16,
        lora_alpha=32,
        lora_dropout=0.05,
        bias="none",
        task_type="CAUSAL_LM",
    )


    trainer = SFTTrainer(
        model,
        train_dataset=dataset,
        data_collator=collator,
        tokenizer=tokenizer,
        peft_config=peft_config,
        dataset_text_field="text",
        max_seq_length=4096,
        args=training_args,

    )

    trainer.train() 

if __name__ == "__main__":
    main()
```

**Comment by lvwerra** [comment]
Hi @ChristianPala, happy new year to you too! Thank you for reporting, we'll look into it! cc @younesbelkada

**Comment by younesbelkada** [comment]
Hi @ChristianPala , happy new year! 
Indeed there was a breaking change, you now need to force-set `gradient_checkpointing=True` when creating your `TrainingArguments` and pass that to `SFTTrainer`. Can you let us know if this works for you?

**Comment by ChristianPala** [comment]
Hello @younesbelkada!

Thank you for the feedback.

Adding the parameter ``gradient_checkpointing=True`` to the ``TrainingArguments`` does not prevent the OOM error on our end, with trl 0.77, below the rest of the setup:

Package                  Version
------------------------ ------------
accelerate               0.25.0
aiohttp                  3.9.1
aiosignal                1.3.1
annotated-types          0.6.0
artificonfig             0.1.2
async-timeout            4.0.3
attrs                    23.2.0
bitsandbytes             0.41.1
certifi                  2023.11.17
cffi                     1.16.0
charset-normalizer       3.3.2
contourpy                1.2.0
cryptography             41.0.7
cycler                   0.12.1
datasets                 2.14.6
dill                     0.3.7
docstring-parser         0.15
einops                   0.7.0
filelock                 3.13.1
flash-attn               2.3.6
fonttools                4.47.0
frozenlist               1.4.1
fsspec                   2023.10.0
huggingface-hub          0.20.1
idna                     3.6
Jinja2                   3.1.2
joblib                   1.3.2
jsonlines                4.0.0
kiwisolver               1.4.5
markdown-it-py           3.0.0
MarkupSafe               2.1.3
matplotlib               3.7.4
mdurl                    0.1.2
models-zoo               0.0.1
mpmath                   1.3.0
multidict                6.0.4
multiprocess             0.70.15
networkx                 3.2.1
ninja                    1.11.1.1
numpy                    1.25.2
nvidia-cublas-cu12       12.1.3.1
nvidia-cuda-cupti-cu12   12.1.105
nvidia-cuda-nvrtc-cu12   12.1.105
nvidia-cuda-runtime-cu12 12.1.105
nvidia-cudnn-cu12        8.9.2.26
nvidia-cufft-cu12        11.0.2.54
nvidia-curand-cu12       10.3.2.106
nvidia-cusolver-cu12     11.4.5.107
nvidia-cusparse-cu12     12.1.0.106
nvidia-nccl-cu12         2.18.1
nvidia-nvjitlink-cu12    12.3.101
nvidia-nvtx-cu12         12.1.105
packaging                23.2
pandas                   2.0.3
peft                     0.7.1
pillow                   10.2.0
pip                      23.3.1
protobuf                 4.25.1
psutil                   5.9.7
pyarrow                  14.0.2
pycparser                2.21
pydantic                 2.5.2
pydantic_core            2.14.5
Pygments                 2.17.2
pyparsing                3.1.1
python-dateutil          2.8.2
python-dotenv            1.0.0
pytz                     2023.3.post1
PyYAML                   6.0.1
regex                    2023.12.25
requests                 2.31.0
rich                     13.7.0
safetensors              0.4.1
scikit-learn             1.3.2
scipy                    1.11.2
setuptools               68.2.2
shtab                    1.6.5
six                      1.16.0
sympy                    1.12
tensorboardX             2.6.2.2
threadpoolctl            3.2.0
tokenizers               0.15.0
torch                    2.1.1
tqdm                     4.66.1
transformers             4.36.2
triton                   2.1.0
trl                      0.7.7
typing_extensions        4.9.0
tyro                     0.6.3
tzdata                   2023.4
urllib3                  2.1.0
wheel                    0.41.2
xxhash                   3.4.1
yarl                     1.9.4

**Comment by younesbelkada** [comment]
Very strange, does the OOM occur with 0.7.2 ? Can you try to pinpoint exactly from which version does this happen so that it will me better help me isolate the bug

**Comment by appleparan** [comment]
@younesbelkada I am experiencing the same issue. The moment I go from 0.7.1 to 0.7.2, memory usage increases sharply, resulting in an OOM (Out of Memory) error.

Here is my requirements.txt
```
accelerate
peft
bitsandbytes
transformers
trl==0.7.2
sentencepiece
huggingface_hub[cli,torch]
scipy
numpy
torch
matplotlib
pandas
scikit-learn
tensorboardX
jupyterlab
ipywidgets
ipykernel
jupyter
python-dotenv
```
